### DISCUSSION TITLE
How to copy all files in a directory?

### DISCUSSION THREAD

**Discussion by nanoqsh** [sagiegurari/cargo-make#663] [Q&A] [answered]
Can I copy new versions of files in specific directory?
For example, I want to build my entire project into a `bin` directory ready for use.

**Comment by sagiegurari** [comment]
you mean copy only newer files? i think i got some duckscript command for that.

**Reply by nanoqsh** [reply]
> you mean copy only newer files?

Yes. And replace old files too, if there is a new version.

**Comment by sagiegurari** [comment, marked as answer]
@nanoqsh  you can use glob_array with is_path_newer to find and copy files.

https://github.com/sagiegurari/duckscript/blob/master/docs/sdk.md#std__fs__GlobArray
https://github.com/sagiegurari/duckscript/blob/master/docs/sdk.md#stdfsispathnewer

so first use glob array to fetch all files recursively, than for/in loop over results and check if newer than target and if so copy.
this is glob copy implementation that you can use as an example:

https://github.com/sagiegurari/duckscript/blob/master/docs/sdk.md#source-13
(sorry for the loooong var names, its like that for internal duckscript commands that are implemented in duckscript to ensure nothing leaks)

**Reply by nanoqsh** [reply]
Thank you

**Comment by nanoqsh** [comment]
So, I think the duckscript is pretty complex within my project. It is a language to be learned. But I found an opportunity to do what I need using the cross platform shell. This is great as it works on all platforms.

But a few questions remain.

I have a project with a server and a client on WASM. I have this project structure:
```
.
├── Makefile.toml
├── server
│   ├── Cargo.toml
│   ├── Makefile.toml
│   └── src
│       └── main.rs
└── web
    ├── Cargo.toml
    ├── Makefile.toml
    ├── src
    │   └── lib.rs
    └── static
        └── index.html
```
I want to build entire project into a `bin` directory. I need to copy `web/static` and the server in `bin`. Also I want to run all build scripts from workspace directory. So Makefile.toml is:
```
[env]
CARGO_MAKE_WORKSPACE_EMULATION = true
CARGO_MAKE_CRATE_WORKSPACE_MEMBERS = [
    "server",
    "web",
]

[tasks.update_bin]
script_runner = "@shell"
script = "mkdir -p bin"
```
server/Makefile.toml:
```
[env]
CARGO_TARGET_DIR = "target/debug" # "target/release" in release build
EXE_SUFFIX = "" # ".exe" on Windows

[tasks.update_bin]
script_runner = "@shell"
script = "cp -u ${CARGO_TARGET_DIR}/server${EXE_SUFFIX} ../bin"
dependencies = ["build"]
```
web/Makefile.toml
```
[tasks.update_bin]
script_runner = "@shell"
script = "cp -ur static ../bin/"
dependencies = ["build"]

[tasks.build]
install_crate = { crate_name = "wasm-pack", binary = "wasm-pack", test_arg = "-V" }
command = "wasm-pack"
args = ["build", "--target", "web", "--out-name", "package", "--dev"]
```
* Can I get the path to the target depending on the build mode? ("target/debug" or "target/release")
* Can I get the full name of the executable depending on the platform? ("server" or "server.exe")

**Reply by nanoqsh** [reply]
Unfortunately this does not work on Windows. I think the most universal approach is to write my own copy in Rust 😔
Perhaps I could get it to work via shell, but it's better to do it perfectly right away.

**Reply by sagiegurari** [reply]
cargo make supports rust based scripts:
https://github.com/sagiegurari/cargo-make#usage-task-command-script-task-examplerust

however, i still think duckscript would be simpler and its cross platform.

**Comment by sagiegurari** [comment]
> Can I get the path to the target depending on the build mode? ("target/debug" or "target/release")

If you run different tasks with release or not (for debug) in the build, you can set an env var in the task.
or you can use profiles also.
for example (profile):

```toml
[env]
TARGET_DIR="target/debug"

[env.release]
TARGET_DIR="target/release"
```

or inside the env of the task that builds
```toml
[env]
TARGET_DIR="target/debug"

[tasks.build-release]
env = { "TARGET_DIR" = "target/release" }
# rest of the task stuff...

[tasks.copy]
#use TARGET_DIR env
```

> Can I get the full name of the executable depending on the platform? ("server" or "server.exe")

Try something like this (untested)

```toml
BINARY_EXTENSION = { source = "${CARGO_MAKE_RUST_TARGET_OS}", default_value = "", mapping = { "windows" = ".exe" } }
BINARY_FILE = "server${BINARY_EXTENSION}
```

**Reply by nanoqsh** [reply]
Thanks!
